### PR TITLE
Add Integration tests with Custom SSL IdentityStore/TrustSore

### PR DESCRIPTION
**Pull request by anpanigr**
Here is the usecase description 
 
Create a MII domain with an attached persistent volume.
 Configure custom identity and custom trust on server template
 Enable SSL on server template with port 8002 (default 7002 does not work) 
 Put the IdentityKeyStore.jks  and TrustKeyStore.jks on /shared directory 
    after administration server pod is started so that it can be accessible 
    from all managed server pods
 Once all servers are started get the JNDI initial context using cluster 
       service URL with t3s protocol.
 Repeat the same after scaling the cluster 

Note - Added a new SslUtils class to deal with SSL related tasks

Jenkin Result
   https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/4413/
   https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/4419/

### CODE REVIEW

File: integration-tests/src/test/resources/wdt-models/mii.ssl.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2020, 2021, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
```
Comment (marinakog) [inline integration-tests/src/test/resources/wdt-models/mii.ssl.yaml:2]: new file, remove 2020

File: integration-tests/src/test/resources/ssl/SslTestClient.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+
+import javax.jms.ConnectionFactory;
+
+public class SslTestClient {
+  private String    url;
+
+  public SslTestClient(String[] args)
```
Comment (marinakog) [inline integration-tests/src/test/resources/ssl/SslTestClient.java:14]: Why do we call it ssl client? there the ssl invoke happen?

**Review by marinakog — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/resources/ssl/SslTestClient.java (hunk 2 above)
Comment (anpanigr) [reply integration-tests/src/test/resources/ssl/SslTestClient.java:14]: It takes t3s URL to get initial context.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                domainUid,
+                domainNamespace,
+                condition.getElapsedTimeInMS(),
+                condition.getRemainingTimeInMS()))
+        .until(domainExists(domainUid, DOMAIN_VERSION, domainNamespace));
+
+    logger.info("Check admin service and pod {0} is created in namespace {1}",
+        adminServerPodName, domainNamespace);
+    checkPodReadyAndServiceExists(adminServerPodName, domainUid, domainNamespace);
+    // Generate JKS Keystore using openssl before
```
Comment (xiancao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java:164]: should be "Generate JKS Keystore using openssl AFTER managed server services and pods are ready"

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                domainNamespace,
+                condition.getElapsedTimeInMS(),
+                condition.getRemainingTimeInMS()))
+        .until(domainExists(domainUid, DOMAIN_VERSION, domainNamespace));
+
+    logger.info("Check admin service and pod {0} is created in namespace {1}",
+        adminServerPodName, domainNamespace);
+    checkPodReadyAndServiceExists(adminServerPodName, domainUid, domainNamespace);
+    // Generate JKS Keystore using openssl before 
+    // managed server services and pods are ready
```
Comment (xiancao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java:165]: Is JKS Keystore required before the managed server service and pods are ready? How do you control the keystore is copied before the managed server service and pods are ready?

File: integration-tests/src/test/resources/bash-scripts/generate-selfsign-jks.sh [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+keytool -importkeystore -destkeystore IdentityKeyStore.jks \
+        -deststorepass changeit -srckeystore identity.p12 \
+        -srcstoretype PKCS12 -srcstorepass changeit 
+
+keytool -import -file cacert.pem -keystore TrustKeyStore.jks \
+        -storepass changeit -noprompt
+
+#keytool -list -v -keystore TrustKeyStore.jks -storepass changeit
+#keytool -list -v -keystore IdentityKeyStore.jks -storepass changeit
```
Comment (xiancao) [inline integration-tests/src/test/resources/bash-scripts/generate-selfsign-jks.sh:37]: remove commented out lines

File: integration-tests/src/test/resources/wdt-models/mii.ssl.yaml (hunk 1 above)
Comment (anpanigr) [reply integration-tests/src/test/resources/wdt-models/mii.ssl.yaml:2]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java (hunk 4 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java:165]: I copy the file immediately after the admin pod is ready and before managed server is being started. That is why insert the copy command between admin server service check and managed server service check.

File: integration-tests/src/test/resources/bash-scripts/generate-selfsign-jks.sh (hunk 5 above)
Comment (anpanigr) [reply integration-tests/src/test/resources/bash-scripts/generate-selfsign-jks.sh:37]: done

**Review by marinakog — APPROVED** [review]

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java (hunk 4 above)
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java:165]: I think there will be possible race condition. If the keystore is required before the managed server is started, then you can start the admin server first, then copy the files and start the managed servers. CheckPodReady() method only checks the status of the pod and does not control when to start the pod.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/SslUtils.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2020, 2021, Oracle and/or its affiliates.
```
Comment (xiancao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/SslUtils.java:1]: remove 2020

File: integration-tests/src/test/resources/ssl/SslTestClient.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  {
+    url      = args[0];
```
Comment (xiancao) [inline integration-tests/src/test/resources/ssl/SslTestClient.java:16]: reformat this line

File: integration-tests/src/test/resources/ssl/SslTestClient.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    try {
+       Context ctx = getInitialContext();
+       System.out.println("Got the Initial JNDI Context ["+ctx+"]");
```
Comment (xiancao) [inline integration-tests/src/test/resources/ssl/SslTestClient.java:19]: add space around "+"

File: integration-tests/src/test/resources/ssl/SslTestClient.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+       String cfName="weblogic.jms.ConnectionFactory";
+       ConnectionFactory qcf=(ConnectionFactory)ctx.lookup(cfName);
+       System.out.println("Looked up default JMS connection factory ["+qcf+"]");
```
Comment (xiancao) [inline integration-tests/src/test/resources/ssl/SslTestClient.java:22]: add space around "+"

File: integration-tests/src/test/resources/ssl/SslTestClient.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+     } catch ( Exception  ex ) { 
+       System.out.println("Got Unknown Exception ["+ ex + "]");
```
Comment (xiancao) [inline integration-tests/src/test/resources/ssl/SslTestClient.java:24]: add space around "+"

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java (hunk 4 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItMiiCustomSslStore.java:165]: After admin server is started it takes few seconds to create the table/JKS files before introspector picks the managed server which takes a minutes to start the server. I also checks if the JMS Server MBean are in proper managed server before proceeding tests. If I see the failure in nightly I have to create a dummy WebLogic Pod to create the table before the domain starts.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/SslUtils.java (hunk 6 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/SslUtils.java:1]: done

File: integration-tests/src/test/resources/ssl/SslTestClient.java (hunk 7 above)
Comment (anpanigr) [reply integration-tests/src/test/resources/ssl/SslTestClient.java:16]: url  = args[0];

File: integration-tests/src/test/resources/ssl/SslTestClient.java (hunk 8 above)
Comment (anpanigr) [reply integration-tests/src/test/resources/ssl/SslTestClient.java:19]: done

File: integration-tests/src/test/resources/ssl/SslTestClient.java (hunk 9 above)
Comment (anpanigr) [reply integration-tests/src/test/resources/ssl/SslTestClient.java:22]: done

File: integration-tests/src/test/resources/ssl/SslTestClient.java (hunk 10 above)
Comment (anpanigr) [reply integration-tests/src/test/resources/ssl/SslTestClient.java:24]: done

**Review by xiancao — APPROVED** [review]

**Pull request merged**